File: 0215.KthLargestElementinanArray.py
from heapq import heappush, heappop

class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        #return sorted(nums)[::-1][k-1]     # Quick Sort: O(NlgN)
        #return self.kthLargest(nums, k, 0, len(nums)-1)    # Divide and Conquer: O(NlgK)
        # Min Heap: O(NlgK)
        heap = []
        for num in nums:
            heappush(heap, num)
            while len(heap) > k:
                heappop(heap)
        return heap[0]

    def kthLargest(self, nums, k, left, right):
        mid = self.partition(nums, left, right)
        if mid == k - 1:
            return nums[mid]
        elif mid > k - 1:
            return self.kthLargest(nums, k, left, mid - 1)
        else:
            return self.kthLargest(nums, k, mid + 1, right)

    def partition(self, nums, left, right):
        pivot = nums[right]
        mid = left - 1
        for i in xrange(left, right):
            if nums[i] > pivot: # Kth Smallest: nums[i] < pivot
                mid += 1
                if i != mid:
                    nums[i], nums[mid] = nums[mid], nums[i]
        mid += 1
        nums[mid], nums[right] = nums[right], nums[mid]
        return mid

File: 0215.KthLargestElementinanArray.go
// ByteDance-Lark
func findKthLargest(nums []int, k int) int {
    /* quick select: 
    average: O(n) + O(n/2) + O(n/4) + O(n/8) + ... = O(2n) = O(n)
    worst case: O(n^2)
    */
    left, right := 0, len(nums)-1
    for left < right {
        mid := partition(nums, left, right)
        if mid == k-1 {
            break
        } else if mid < k-1 {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return nums[k-1]
}

func partition(nums []int, l, r int) int {
    mid := l
    for i := l; i < r; i++ {
        if nums[i] > nums[r] {
            nums[mid], nums[i] = nums[i], nums[mid]
            mid++
        }
    }
    nums[mid], nums[r] = nums[r], nums[mid]
    return mid
}

/*
func partition(nums []int, l, r int) int {
    pivot := nums[l]
    for l < r {
        for ; l < r && nums[r] <= pivot; r-- {}
        nums[l] = nums[r]
        for ; l < r && nums[l] > pivot; l++ {}
        nums[r] = nums[l]
    }
    nums[l] = pivot
    return l
}
*/